pkg/syncer/client/client.go

```diff
@@ -197,33 +202,28 @@ func (c *Client) update(ctx context.Context, obj client.Object, updateFn update,
 	return nil, status.ResourceWrap(lastErr, "exceeded max tries to update", obj)
 }
 
-// Upsert creates or updates the given obj in the Kubernetes cluster and records prometheus metrics.
-// This operation always involves retrieving the resource from API Server before actually creating or updating it.
-func (c *Client) Upsert(ctx context.Context, obj client.Object) status.Error {
+// Update posts the update to the server and records latency metrics.
+// Specify a ResourceVersion to avoid overwriting asynchronous changes.
+func (c *Client) Update(ctx context.Context, obj client.Object) status.Error {
 	description := getResourceInfo(obj)
-	namespacedName := getNamespacedName(obj)
-
-	// We don't actually care about the object on the cluster, we just want to
-	// check if it exists before deciding whether to create or update it.
-	u := &unstructured.Unstructured{}
-	u.SetGroupVersionKind(obj.GetObjectKind().GroupVersionKind())
-	if err := c.Client.Get(ctx, namespacedName, u); err != nil {
-		if apierrors.IsNotFound(err) {
-			return c.Create(ctx, obj)
-		}
-		return status.ResourceWrap(err, "failed to get resource for update", obj)
-	}
-
 	klog.V(1).Infof("Will update %s to %s", description, spew.Sdump(obj))
+	oldV := resourceVersion(obj)
 	start := time.Now()
 	err := c.Client.Update(ctx, obj)
 	c.recordLatency(start, "update", obj.GetObjectKind().GroupVersionKind().Kind, metrics.StatusLabel(err))
 	m.RecordAPICallDuration(ctx, "update", m.StatusTagKey(err), obj.GetObjectKind().GroupVersionKind(), start)
-
-	if err != nil {
+	switch {
+	case apierrors.IsNotFound(err):
+		return ConflictUpdateDoesNotExist(err, obj)
```

Comment:
`c.Client.Apply` returns `ConflictUpdateDoesNotExist` error directly, no need to wrap it again.

Reply:
There is no c.Client.Apply. You're confusing the upstream client library with the syncer package functions.

Reply:
Ahh, I see. That's the k8s client.